src/data-fetchers/bam/bam-worker.ts

```diff
@@ -285,129 +281,122 @@ const tilesetInfo = (uid: string) => {
-                        .then(records => {
-                            const mappedRecords = records.map(rec =>
-                                bamRecordToJson(rec, chromName, cumPositions[i].pos)
-                            );
-                            tileValues.set(
-                                `${uid}.${z}.${x}`,
-                                (tileValues.get(`${uid}.${z}.${x}`) as JsonBamRecord[]).concat(mappedRecords)
-                            );
-                            return [];
-                            // return mappedRecords;
-                        })
-                );
-
-                // continue onto the next chromosome
-                minX = chromEnd;
-            } else {
-                const startPos = Math.floor(minX - chromStart);
-                const endPos = Math.ceil(maxX - chromStart);
-                // the end of the region is within this chromosome
-                recordPromises.push(
-                    bam.file.getRecordsForRange(chromName, startPos, endPos, opt).then(records => {
-                        const mappedRecords = records.map(rec => bamRecordToJson(rec, chromName, cumPositions[i].pos));
-                        tileValues.set(
-                            `${uid}.${z}.${x}`,
-                            (tileValues.get(`${uid}.${z}.${x}`) as JsonBamRecord[]).concat(mappedRecords)
-                        );
-                        return [];
-                    })
-                );
-                // end the loop because we've retrieved the last chromosome
-                break;
-            }
+    const regions: [{ chr: string; pos: number }, RelativeRegion][] = [];
+
+    for (const c of chromInfo.cumPositions) {
+        const extent = c.pos + chromInfo.chromLengths[c.chr];
+
+        // start of visible region
+        if (!(c.pos <= maxX && minX < extent)) {
+            continue;
+        }
+
+        if (maxX > extent) {
+            // the visible region extends beyond the end of this chromosome
+            // fetch from the start until the end of the chromosome
+            const region = { start: minX - c.pos, end: extent - c.pos };
+            regions.push([c, region]);
+
+            // continue onto the next chromosome
+            minX = extent;
+        } else {
+            // the end of the region is within this chromosome
+            const region = { start: Math.floor(minX - c.pos), end: Math.ceil(maxX - c.pos) };
+            regions.push([c, region]);
+
+            // end the loop because we've retrieved the last chromosome
+            break;
         }
     }
 
-    // flatten the array of promises so that it looks like we're getting one long list of value
-    return Promise.all(recordPromises).then(values => {
-        return values.flat();
+    return regions;
+}
+
+const tile = async (uid: string, z: number, x: number): Promise<JsonBamRecord[]> => {
+    const tileCacheKey = `${uid}.${z}.${x}`;
```

Comment:
I'm not sure if this cache key is all that useful. The way that bam works there is a finite set of records at a single resolution. This caching strategy will redundantly cache records which we have already founds at a different zoom level. 

Ideally we could have some sort of interval cache, where if the interval (region) has already been visited, we know we have all the records for that region.